From e16791cac4c90e6283237d2884b5a48738c96019 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Wed, 30 Mar 2011 06:55:49 -0600
Subject: [PATCH 031/177] - added flags to c2_md_store_locte() to handle open
 case that needs to locate a cob not on the store but rather in list of opened
 files.

---
 cob/cob.h           |  2 +-
 mdservice/md_foms.c | 43 ++++++++++++++++++++++++-------------------
 mdstore/mdstore.c   | 12 ++++++++++--
 mdstore/mdstore.h   |  8 +++++++-
 4 files changed, 42 insertions(+), 23 deletions(-)

diff --git a/cob/cob.h b/cob/cob.h
index 5bb3380..b1bce21 100644
--- a/cob/cob.h
+++ b/cob/cob.h
@@ -238,7 +238,7 @@ struct c2_cob {
 /** 
    Cob flags and valid attributes 
 */
-enum ca_valid {
+enum c2_cob_ca_valid {
         CA_NSKEY      = (1 << 0),  /**< nskey in cob is up-to-date */
         CA_NSKEY_FREE = (1 << 1),  /**< cob responsible for dealloc of nskey */
         CA_NSKEY_DB   = (1 << 2),  /**< db responsible for dealloc of nskey */
diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index a5c8037..ba0309e 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -158,8 +158,9 @@ static int c2_md_fom_link_state(struct c2_fom *fom)
 
                 c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, 
-                                        &cob, &ctx->ft_tx->tx_dbtx);
+                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                        C2_MD_STORE_LOCATE_STORE,
+                                        &ctx->ft_tx->tx_dbtx);
                 if (rc)
                         goto out;
 
@@ -219,8 +220,9 @@ static int c2_md_fom_unlink_state(struct c2_fom *fom)
 
                 c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, 
-                                        &cob, &ctx->ft_tx->tx_dbtx);
+                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                        C2_MD_STORE_LOCATE_STORE,
+                                        &ctx->ft_tx->tx_dbtx);
                 if (rc)
                         goto out;
 
@@ -280,8 +282,9 @@ static int c2_md_fom_rename_state(struct c2_fom *fom)
 
                 c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, 
-                                        &cob, &ctx->ft_tx->tx_dbtx);
+                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                        C2_MD_STORE_LOCATE_STORE,
+                                        &ctx->ft_tx->tx_dbtx);
                 if (rc)
                         goto out;
 
@@ -341,8 +344,9 @@ static int c2_md_fom_open_state(struct c2_fom *fom)
 
                 c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, 
-                                        &cob, &ctx->ft_tx->tx_dbtx);
+                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                        C2_MD_STORE_LOCATE_STORE,
+                                        &ctx->ft_tx->tx_dbtx);
                 if (rc == -ENOENT) {
                         /*
                          * @todo: create file if open mode says so.
@@ -397,11 +401,9 @@ static int c2_md_fom_close_state(struct c2_fom *fom)
 
                 c2_md_make_fid(&fid, &body->b_tfid);
 
-#if 0
-                /** @todo: Find a cob in open files table. */
-                rc = c2_md_store_opened(site->s_mdstore, &fid, 
-                                        &cob, &ctx->ft_tx->tx_dbtx);
-#endif
+                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                        C2_MD_STORE_LOCATE_OPEN,
+                                        &ctx->ft_tx->tx_dbtx);
                 if (rc)
                         goto out;
                 rep = c2_fop_data(fop_rep);
@@ -460,8 +462,9 @@ static int c2_md_fom_setattr_state(struct c2_fom *fom)
 
                 c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, 
-                                        &cob, &ctx->ft_tx->tx_dbtx);
+                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                        C2_MD_STORE_LOCATE_STORE,
+                                        &ctx->ft_tx->tx_dbtx);
                 if (rc)
                         goto out;
 
@@ -521,8 +524,9 @@ static int c2_md_fom_getattr_state(struct c2_fom *fom)
 
                 c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, 
-                                        &cob, &ctx->ft_tx->tx_dbtx);
+                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                        C2_MD_STORE_LOCATE_STORE,
+                                        &ctx->ft_tx->tx_dbtx);
                 if (rc)
                         goto out;
 
@@ -582,8 +586,9 @@ static int c2_md_fom_readdir_state(struct c2_fom *fom)
 
                 c2_md_make_fid(&fid, &body->b_tfid);
 
-                rc = c2_md_store_locate(site->s_mdstore, &fid, 
-                                        &cob, &ctx->ft_tx->tx_dbtx);
+                rc = c2_md_store_locate(site->s_mdstore, &fid, &cob, 
+                                        C2_MD_STORE_LOCATE_STORE,
+                                        &ctx->ft_tx->tx_dbtx);
                 if (rc)
                         goto out;
 
diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 3e189f2..6fd5ff6 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -432,13 +432,21 @@ static void c2_md_store_make_nskey(struct c2_cob_nskey **keyh,
    Find cob by fid.
 */
 int c2_md_store_locate(struct c2_md_store *md, struct c2_fid *fid,
-                       struct c2_cob **cob, struct c2_db_tx *tx)
+                       struct c2_cob **cob, int flags, struct c2_db_tx *tx)
 {
         struct c2_cob_oikey oikey;
         int                 rc;
 
         c2_md_store_make_oikey(&oikey, fid);
-        rc = c2_cob_locate(&md->md_dom, &oikey, cob, tx);
+
+        if (flags == C2_MD_STORE_LOCATE_STORE) {
+                rc = c2_cob_locate(&md->md_dom, &oikey, cob, tx);
+        } else {
+                /*
+                 * @todo: locate cob in opened cobs table.
+                 */
+                rc = -EOPNOTSUPP;
+        }
 
         return rc;
 }
diff --git a/mdstore/mdstore.h b/mdstore/mdstore.h
index 83e4e3b..b9fc47b 100644
--- a/mdstore/mdstore.h
+++ b/mdstore/mdstore.h
@@ -18,6 +18,12 @@ struct c2_md_store {
         struct c2_fid         md_rootid;
 };
 
+enum c2_md_store_locate_flags {
+        C2_MD_STORE_LOCATE_STORE   = 1 << 0,
+        C2_MD_STORE_LOCATE_OPEN    = 1 << 1,
+        C2_MD_STORE_LOCATE_ORPHANS = 1 << 2
+};
+
 /**
    Initialize mdstore with passed cob comain id and database env.
 */
@@ -150,7 +156,7 @@ int c2_md_store_readdir(struct c2_md_store *md,
    Find cob by fid.
 */
 int c2_md_store_locate(struct c2_md_store *md, struct c2_fid *fid,
-                       struct c2_cob **cob, struct c2_db_tx *tx);
+                       struct c2_cob **cob, int flags, struct c2_db_tx *tx);
 
 /**
    Find cob by parent and name.
-- 
1.8.3.2

